enhydris_autoprocess/models.py

```diff
@@ -1,365 +0,0 @@
-import csv
-import datetime as dt
-import re
-from io import StringIO
-
-from django.db import IntegrityError, models
-from django.utils.translation import gettext_lazy as _
-
-import numpy as np
-import pandas as pd
-from haggregate import aggregate, regularize
-
-from enhydris.models import Timeseries, TimeseriesGroup
-
-from . import tasks
-
-
-class AutoProcess(models.Model):
-    timeseries_group = models.ForeignKey(TimeseriesGroup, on_delete=models.CASCADE)
-
-    class Meta:
-        verbose_name_plural = _("Auto processes")
-
-    def execute(self):
-        self.htimeseries = self._subclass.source_timeseries.get_data(
-            start_date=self._get_start_date()
-        )
-        result = self._subclass.process_timeseries()
-        self._subclass.target_timeseries.append_data(result)
-
-    @property
-    def _subclass(self):
-        """Return the AutoProcess subclass for this instance.
-
-        AutoProcess is essentially an abstract base class; its instances are always one
-        of its subclasses, i.e. Checks, CurveInterpolation, or Aggregation. Sometimes we
-        might have an AutoProcess instance without yet knowing what subclass it is; in
-        that case, "myinstance._subclass" is the subclass.
-        """
-        for alternative in ("checks", "curveinterpolation", "aggregation"):
-            if hasattr(self, alternative):
-                result = getattr(self, alternative)
-                if hasattr(self, "htimeseries"):
-                    result.htimeseries = self.htimeseries
-                return result
-
-    def _get_start_date(self):
-        start_date = self._subclass.target_timeseries.end_date
-        if start_date:
-            start_date += dt.timedelta(minutes=1)
-        return start_date
-
-    def save(self, *args, **kwargs):
-        result = super().save(*args, **kwargs)
-        self._subclass._check_integrity()
-
-        # We delay running the task by one second; otherwise it may run before the
-        # current transaction has been committed.
-        tasks.execute_auto_process.apply_async(args=[self.id], countdown=1)
```

Comment:
You can use `transaction.on_commit` to avoid non-deterministic behavior like so:
```python
from django.db import transaction
transaction.on_commit(lambda: tasks.execute_auto_process.apply_async(args=[self.id]))
```
This is important since if at one time the database's connection was slow, this will result in a race condition.

Note: You'll need `TransactionTestCase` for tests that tested that the task is fired.

Reply:
The problem with `on_commit` is that it won't give me the time series id. So the solution could be to add the time series id to a thread local list, and then use `on_commit` to execute `apply_async`, which will get the time series id from the list. But, if you don't object, I think this would be overkill and just delaying for one second is fine I think. And if the database happens one day to be unbelievably slow, this won't really do much harm—celery will have an internal error, and the auto process won't be executed (but it will be executed the next time data is uploaded).

Reply:
Added a commit that fixes that.